Spring cleaning: Apply clang-tidy to src/widget.

rryan opened the pull request:
Using LLVM 3.8.0:

$ clang-tidy --fix --checks="*,-llvm-header-guard,-readability-implicit-bool-cast,-readability-else-after-return,-google-readability-todo" $(git ls src/widget | xargs) -- (flags to compile)

rryan committed:
Spring cleaning: Apply clang-tidy to src/widget.

Using LLVM 3.8.0:

$ clang-tidy --fix --checks="*,-llvm-header-guard,-readability-implicit-bool-cast,-readability-else-after-return,-google-readability-todo" $(git ls src/widget | xargs) -- (flags to compile)

ywwg commented on src/widget/wlibrarysidebar.cpp:
```diff
@@ -196,7 +195,7 @@ void WLibrarySidebar::keyPressEvent(QKeyEvent* event) {
 
 void WLibrarySidebar::selectIndex(const QModelIndex& index) {
-    QItemSelectionModel* pModel = new QItemSelectionModel(model());
+    auto  pModel = new QItemSelectionModel(model());
```
it's weird that it puts a space after "auto" everywhere

rryan replied:
Yea -- not sure why it did that...

daschuer replied:
It looks like the "auto" replacement expects the a 
`QItemSelectionModel *pModel` syntax where it is correct to add a space after *

ywwg replied:
does that mean we need to make this auto* pBlah?

daschuer replied:
No, both is allowed. I think I have just found a reason for the double space. 
clang-tidy replaces `"QItemSelectionModel *"` by `"auto "` regardless if there is already a space between the * and the variable name or not.

daschuer commented on src/widget/controlwidgetconnection.cpp:
```diff
@@ -15,27 +15,26 @@ ControlWidgetConnection::ControlWidgetConnection(WBaseWidget* pBaseWidget,
 
-ControlWidgetConnection::~ControlWidgetConnection() {
-}
+ControlWidgetConnection::~ControlWidgetConnection() = default;
```
This is redundant in many ways. It overrides the already empty destructor of the base class, so it should be fine to delete this together with the prototype in the header file. 

In case of a destructor, the default keyword is always { }.  So there is IMHO no benefit of using it, it is longer, less explicit and requires C++11 knowledge to understand.  The { } notation has also the benefit that it provides a skeleton that can be filled if required. 

By the way: For the constructor case, the default keyword is very nice.

rryan replied:
I went ahead and deleted the default constructors entirely for this PR

daschuer commented on src/widget/slidereventhandler.h:
```diff
@@ -105,7 +105,7 @@ class SliderEventHandler {
     void wheelEvent(T* pWidget, QWheelEvent* e) {
         // For legacy (MIDI) reasons this is tuned to 127.
-        double wheelAdjustment = ((QWheelEvent *)e)->delta() / (120.0 * 127.0);
+        double wheelAdjustment = (e)->delta() / (120.0 * 127.0);
```
redundant parentheses (e)

daschuer commented on src/widget/wcoverartmenu.cpp:
```diff
@@ -57,8 +57,8 @@ void WCoverArtMenu::slotChange() {
     QStringList extensions = CoverArtUtils::supportedCoverArtExtensions();
-    for (QStringList::iterator it = extensions.begin(); it != extensions.end(); ++it) {
-        it->prepend("*.");
+    for (auto & extension : extensions) {
```
`const auto&` or `auto&&`

daschuer commented on src/widget/wbattery.h:
```diff
@@ -16,8 +16,8 @@ class Battery;
   public:
-    WBattery(QWidget* parent=NULL);
-    virtual ~WBattery();
+    explicit WBattery(QWidget* parent=nullptr);
```
`parent = nullptr` with spaces

ywwg replied:
I think this tidy pass is not redoing all of our whitespace at all, that would be a different and much bigger diff.  It's something we should definitely try to do, though!

daschuer commented on src/widget/wcoverartmenu.h:
```diff
@@ -17,8 +17,8 @@
   public:
-    WCoverArtMenu(QWidget *parent = 0);
-    virtual ~WCoverArtMenu();
+    explicit WCoverArtMenu(QWidget *parent = nullptr);
```
* aligned with the type

daschuer commented on src/widget/wdisplay.h:
```diff
@@ -30,15 +30,15 @@
   public:
-    WDisplay(QWidget *parent=NULL);
-    virtual ~WDisplay();
+    explicit WDisplay(QWidget *parent=nullptr);
```
QWidget* parent = nullptr

daschuer commented on src/widget/woverview.cpp:
```diff
@@ -343,12 +343,10 @@ void WOverview::paintEvent(QPaintEvent *) {
-            for (unsigned int i = 0; i < m_markRanges.size(); ++i) {
-                WaveformMarkRange& currentMarkRange = m_markRanges[i];
-
+            for (auto & currentMarkRange : m_markRanges) {
```
auto&&

daschuer commented on src/widget/woverviewhsv.cpp:
```diff
@@ -58,7 +58,7 @@ bool WOverviewHSV::drawNextPixmapPart() {
 
     QPainter painter(m_pWaveformSourceImage);
-    painter.translate(0.0,(double)m_pWaveformSourceImage->height()/2.0);
+    painter.translate(0.0,static_cast<double>(m_pWaveformSourceImage->height())/2.0);
```
spaces

daschuer commented on src/widget/woverviewlmh.cpp:
```diff
@@ -60,7 +60,7 @@ bool WOverviewLMH::drawNextPixmapPart() {
 
     QPainter painter(m_pWaveformSourceImage);
-    painter.translate(0.0,(double)m_pWaveformSourceImage->height()/2.0);
+    painter.translate(0.0,static_cast<double>(m_pWaveformSourceImage->height())/2.0);
```
spaces

daschuer commented on src/widget/woverviewrgb.cpp:
```diff
@@ -56,7 +56,7 @@ bool WOverviewRGB::drawNextPixmapPart() {
 
     QPainter painter(m_pWaveformSourceImage);
-    painter.translate(0.0,(double)m_pWaveformSourceImage->height()/2.0);
+    painter.translate(0.0,static_cast<double>(m_pWaveformSourceImage->height())/2.0);
```
spaces

daschuer commented on src/widget/wspinny.cpp:
```diff
@@ -106,17 +106,17 @@ void WSpinny::onVinylSignalQualityUpdate(const VinylSignalQualityReport& report)
 
     for (int y = 0; y < m_iVinylScopeSize; ++y) {
-        QRgb *line = (QRgb *)m_qImage.scanLine(y);
+        QRgb *line = reinterpret_cast<QRgb *>(m_qImage.scanLine(y));
```
reinterpret_cast, really?

rryan replied:
scanLine returns a uchar* so I believe this is the correct cast, no?

daschuer replied:
Ah yes, I see: http://doc.qt.io/qt-4.8/qimage.html#scanLine

daschuer commented:
Nice, was this done all automatic? 
I have added some comments and some style nits, that where already there before.

rryan commented:
> Nice, was this done all automatic? 
> I have added some comments and some style nits, that where already there before.

Yep! clang-tidy is a modernization tool, not a formatting tool so the formatting isn't going to be perfect. It's meant to be used with clang-format to keep the style as per your rules but without a clang-format rule it mostly leaves things as is. As you and Owen pointed out it does odd things in certain places. I'd prefer to not focus on making this perfect as per our style guide and instead focus on the improvements it can bring relative to the status quo.

daschuer commented:
Ignore the style comments if you like, it was just too eye catching.

rryan committed:
Fix some clang-tidy-produced lint and convert foreach loops to range-based for loops.

rryan commented:
I cleaned up the extra spaces it inserted, converted auto& to auto&& throughout, and converted foreach's to range-based fors (manually, unfortunately. though clang-tidy is pluggable so we could build a tidy that converts it :) but that would probably take longer than manually fixing all the foreach's).

ywwg commented on src/widget/weffectpushbutton.cpp:
```diff
@@ -64,7 +64,7 @@ void WEffectPushButton::mousePressEvent(QMouseEvent* e) {
     // QAction.
     double leftValue = getControlParameterLeft();
-    foreach (QAction* action, m_pButtonMenu->actions()) {
+    for (const auto& action : m_pButtonMenu->actions()) {
```
so this is a constant reference to a pointer?  That seems poor, "for (auto* action :" would be the preferred way to do it I think.

rryan replied:
Derp. I did this manually (since it was a foreach). Replaced with QAction* for readability.

ywwg commented:
Just one issue I see with odd range-based for loops that generate constant references to pointers.  Otherwise, LGTM

rryan committed:
Replace const-reference-to-pointer with by-value copy.

rryan committed:
Delete default destructors.

daschuer commented:
LGTM, thank you.

rryan commented:
Thanks for the reviews!
